sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1445,45 +1494,108 @@ _pi_queue::pi_queue_group_t::getZeQueue(uint32_t *QueueGroupOrdinal) {
 
   *QueueGroupOrdinal = Queue->Device->QueueGroup[QueueType].ZeOrdinal;
-
-  ze_command_queue_handle_t &ZeQueue = ZeQueues[CurrentIndex];
-  if (ZeQueue)
-    return ZeQueue;
-
-  // Command queue is not available at chosen index. So we create it below.
-  ZeStruct<ze_command_queue_desc_t> ZeCommandQueueDesc;
-
-  // Adjust the index to the L0 queue group since we represent "main" and "link"
+  // Adjust the index to the L0 queue group since we represent "main" and
+  // "link"
   // L0 groups with a single copy group ("main" would take "0" index).
   auto ZeCommandQueueIndex = CurrentIndex;
   if (QueueType == queue_type::LinkCopy && Queue->Device->hasMainCopyEngine()) {
     ZeCommandQueueIndex -= 1;
   }
+  *QueueIndex = ZeCommandQueueIndex;
 
-  ZeCommandQueueDesc.ordinal = *QueueGroupOrdinal;
-  ZeCommandQueueDesc.index = ZeCommandQueueIndex;
+  return CurrentIndex;
+}
+
+// Return a queue descriptor using the provided ordinal and index.
+void _pi_queue::pi_queue_group_t::createQueueDesc(
+    uint32_t Index, uint32_t Ordinal,
+    ZeStruct<ze_command_queue_desc_t> &ZeCommandQueueDesc) {
+
+  ZeCommandQueueDesc.ordinal = Ordinal;
+  ZeCommandQueueDesc.index = Index;
   ZeCommandQueueDesc.mode = ZE_COMMAND_QUEUE_MODE_ASYNCHRONOUS;
 
   // Evaluate performance of explicit usage for "0" index.
-  if (ZeCommandQueueIndex != 0) {
+  if (Index != 0) {
     ZeCommandQueueDesc.flags = ZE_COMMAND_QUEUE_FLAG_EXPLICIT_ONLY;
   }
 
   zePrint("[getZeQueue]: create queue ordinal = %d, index = %d "
           "(round robin in [%d, %d])\n",
           ZeCommandQueueDesc.ordinal, ZeCommandQueueDesc.index, LowerIndex,
           UpperIndex);
+}
+
+// This function will return one of possibly multiple available native
+// queues and the value of the queue group ordinal.
+ze_command_queue_handle_t &
+_pi_queue::pi_queue_group_t::getZeQueue(uint32_t *QueueGroupOrdinal) {
+
+  // QueueIndex is the proper L0 index.
+  // Index is the plugins concept of index, with main and link copy engines in
+  // one range.
+  uint32_t QueueIndex;
+  auto Index = getQueueIndex(QueueGroupOrdinal, &QueueIndex);
+
+  ze_command_queue_handle_t &ZeQueue = ZeQueues[Index];
+  if (ZeQueue)
+    return ZeQueue;
+
+  ZeStruct<ze_command_queue_desc_t> ZeCommandQueueDesc;
+  createQueueDesc(QueueIndex, *QueueGroupOrdinal, ZeCommandQueueDesc);
 
   auto ZeResult = ZE_CALL_NOCHECK(
       zeCommandQueueCreate, (Queue->Context->ZeContext, Queue->Device->ZeDevice,
                              &ZeCommandQueueDesc, &ZeQueue));
   if (ZeResult) {
     die("[L0] getZeQueue: failed to create queue");
   }
+
   return ZeQueue;
 }
 
+// This function will return one of possibly multiple available
+// immediate commandlists associated with this Queue.
+pi_command_list_ptr_t &
+_pi_queue::pi_queue_group_t::getImmCmdList(bool UseCopyEngine) {
+
+  uint32_t QueueIndex, QueueOrdinal;
+  auto Index = Queue->getQueueGroup(UseCopyEngine)
+                   .getQueueIndex(&QueueOrdinal, &QueueIndex);
```

Comment:
This will still create multiple immediate command-list to multiple compute engines, that is an unwanted overhead (and is not working stable currently)

Reply:
The longer-term direction is to use all compute engines so it is better to implement it that way right now, rather than needing another change later. We have an env var to restrict usage to index 0. We should use that while the L0 bug exists.

Reply:
Please make it using "0" index only. I am fine if it will be around where the allowed engines are filtered.

Reply:
I don't understand what you mean by " I am fine if it will be around where the allowed engines are filtered.".
Do you mean leave the code to potentially use all engines, but force a 0 now?

Reply:
Yes. Here: https://github.com/intel/llvm/pull/5833/files#r843258684